system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -833,6 +848,346 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+                else
+                  tgtXpathCtx->addInputValue(name, queryAttrValue(node,"value"));
+            }
+        }
+        xmlXPathFreeObject(eval);
+    }
+    const char *getXPathString(const char *xpath, StringBuffer &s) const override
+    {
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) xpath, xpathCtx);
+        if (eval)
+        {
+            xmlChar *v = xmlXPathCastToString(eval);
+            xmlXPathFreeObject(eval);
+            if (v)
+            {
+                if (*v)
+                    s.append((const char *)v);
+                xmlFree(v);
+                return s;
+            }
+        }
+        return nullptr;
+    }
+    __int64 getXPathInt64(const char *xpath, __int64 dft=0) const override
+    {
+        __int64 ret = dft;
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) xpath, xpathCtx);
+        if (eval)
+        {
+            xmlChar *val = xmlXPathCastToString(eval);
+            xmlXPathFreeObject(eval);
+            if (val)
+            {
+                if (*val)
+                    ret = _atoi64((const char *)val);
+                xmlFree(val);
+            }
+        }
+        return ret;
+    }
+    bool getXPathBool(const char *xpath, bool dft=false) const override
+    {
+        bool ret = dft;
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) xpath, xpathCtx);
+        if (eval)
+        {
+            xmlChar *val = xmlXPathCastToString(eval);
+            xmlXPathFreeObject(eval);
+            if (val)
+            {
+                if (*val)
+                    ret = strToBool((const char *)val);
+                xmlFree(val);
+            }
+        }
+        return ret;
+    }
+
+    xmlNodePtr getSection(const char *name)
+    {
+        return getSectionNode(name, nullptr);
+    }
+    xmlNodePtr ensureSection(const char *name)
+    {
+        xmlNodePtr sect = getSection(name);
+        if (sect)
+            return sect;
+        return xmlNewChild(root, nullptr, (const xmlChar *) name, nullptr);
+    }
+    void removeSection(const char *name)
+    {
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) name, xpathCtx);
+        if (!eval)
+            return;
+        if (XPATH_NODESET==eval->type && !xmlXPathNodeSetIsEmpty(eval->nodesetval))
+        {
+            xmlNodeSetPtr ns = eval->nodesetval;
+            int count = xmlXPathNodeSetGetLength(ns);
+            for(int i=0; i<count; i++)
+            {
+                xmlNodePtr node = xmlXPathNodeSetItem(ns, i);
+                if (node)
+                {
+                    xmlUnlinkNode(node);
+                    xmlFreeNode(node);
+                    ns->nodeTab[i]=nullptr;
+                }
+            }
+        }
+        xmlXPathFreeObject(eval);
+    }
+    xmlNodePtr replaceSection(const char *name)
+    {
+        removeSection(name);
+        return xmlNewChild(root, nullptr, (const xmlChar *) name, nullptr);
```

Comment:
Should there be verification that 'name' is an element name and not a path?

Reply:
If setContent is called with an XPath and not a section name, replaceSection will act on the XPath by allowing removeSection to remove all matching occurrences. The new child of 'root' might not be created, but we've already potentially destroyed the document.

Whether it's here or in setContent, it seems like we need to verify that we're manipulating a section and not an arbitrary location. Otherwise it seems like a loophole waiting to be abused.

Reply:
I hope that people who knowingly try to pass in an xpath for the name get what they deserve when the document is destroyed.  But I do now check name is not a path via the call to sanityCheckSectionName().